### ISSUE TITLE
Why do we not autogenerate traits api docs?

### ISSUE THREAD

**Issue by rahulporuri** [enthought/traits#1243]
At the moment, it looks like we are storing the API docs in the codebase instead of dynamically generating it. The API reference is stored in https://github.com/enthought/traits/tree/master/docs/source/traits_api_reference . Is there any specific reason why we are not dynamically generating the API docs every time?

**Comment by mdickinson** [comment]
> Is there any specific reason why we are not dynamically generating the API docs every time?

I assume you're talking about using `sphinx-apidoc`? (We're already using autodoc, so already doing some autogeneration.)

If so, then yes, there is a reason: out of the box, the `sphinx-apidoc` results are ugly and contain unnecessary extra levels of nesting, and I haven't yet invested time to find a way to both (a) autogenerate and (b) produce an acceptable level of non-ugliness. PRs that introduce autogeneration and don't produce output worse than the current output are welcome.

**Comment by mdickinson** [comment]
To see what I mean, take a look at the Envisage docs, which _are_ generated using `sphinx-apidoc`: https://docs.enthought.com/envisage/api/modules.html

The nesting of everything under "envisage package" is redundant, and there are spurious "Subpackages", "Submodules" and "Module Contents" links all over the place.

In the case of Envisage, this is better than nothing (which is what we had before).

So that answers the question in the issue title. Should we close, or do you want to repurpose this issue as a "Let's autogenerate the documentation" one?

**Comment by rahulporuri** [comment]
>  Should we close

So, just to confirm, at the moment, if i were to add a new module, a new class or a new method, I would manually need to use apidoc to generate the relevant API docs? Or is there a command somewhere that I am not familiar with. Or is are the API docs updated manually?

I don't know if that is documented somewhere and if it is, then i'm happy with closing this issue. If not, can we change this issue to address that issue?

**Comment by mdickinson** [comment]
> So, just to confirm, at the moment, if i were to add a new module, a new class or a new method, I would manually need to use apidoc to generate the relevant API docs?

Yes, that's one way, or you could just add the `.rst` file by hand and use `automodule` at the top of it. That way it ends up being just 10 or so lines. (This is what I've done for Traits Futures.)

I'm entirely open to autogenerated documentation, but not if it's even worse quality-wise that what we have now. And it's non-trivial effort to do that, which is why it hasn't happened yet.

**Comment by mdickinson** [comment]
Traits Futures example: https://github.com/enthought/traits-futures/blob/master/docs/source/api/traits_futures.background_progress.rst

**Comment by mdickinson** [comment]
> If not, can we change this issue to address that issue?

I retitled the issue.

**Comment by mdickinson** [comment]
Note that more recent versions of Sphinx allow specification of your own templates for sphinx-apidoc: https://www.sphinx-doc.org/en/master/man/sphinx-apidoc.html#cmdoption-sphinx-apidoc-t

That's probably what we need to look at to get reasonable output.

**Comment by mdickinson** [comment]
Using the templates turned out to be fairly straightforward (though it took some tweaking to get them exactly right): here's what worked on Traits Futures: enthought/traits-futures#181.

**Comment by kitchoi** [comment]
In addition to the redundant "subpackages" and "submodules" links...

Actually we are already autogenerating documentation via the use of `automodule`, we are just not auto generating the source file that includes these `automodule` (kind of meta...).

This is related to #1213: Perhaps we should also consider defining `__all__` in modules for documentation purposes. Otherwise all classes and functions with "public" names are going to be picked up the automatic tool.

This goes together with #1213. Having stated stuff outside of `traits.api` are not part of the public API is good. But I am not this is a common convention in the community. I am a typical user who doesn't read documentation back to back, so I will likely miss that statement and assume anything documented was part of the public API.  It would be good if the actual public "API References" doesn't include all the bits that are not meant to be consumed by the public.

Alternative to automatic doc generation, we could keep maintaining the API documentation source manually: Each time when someone adds a new function or class via a PR, they will need to decide whether it needs to be added to the API doc as well. Like this PR: https://github.com/enthought/traits/pull/1071/files

The slight downside of this approach is that it is more likely to create PR merge conflicts, but they tend to be trivial and manageable.

**Comment by mdickinson** [comment]
> It would be good if the actual public "API References" doesn't include all the bits that are not meant to be consumed by the public.

I think we still want all this documentation built and visible. I know I find it useful. But we could organize the documentation to put it under a "Developer Reference" heading, or otherwise make it clear that it's intended for developers.
